### PR TITLE
[TACHYON-424] S3 client v2

### PR DESCRIPTION
**Pull request by calvinjia**
Rebased version of #891 

https://tachyon.atlassian.net/projects/TACHYON/issues/TACHYON-424

This PR implements an S3 Client using the jets3t library.

A new s3 submodule under /underfs contains the logic for the client. The user may enable this client as their s3 client by removing the s3n:// prefix from the hadoop prefixes in their configuration.

The implementation conforms to our UnderFileSystem interface and unit tests. It also passes all applicable core tests when using the S3UnderFileSystemCluster. Since the S3UnderFileSystem is implemented as a real S3 backend, the tests are disabled by default.

### CODE REVIEW

**Comment by calvinjia** [comment]
@apc999 Thanks for the feedback, I've updated the code accordingly.

**Comment by aarondav** [comment]
Very high level point, but out of curiosity, why was jets3t used as the S3 library over the AWS SDK? There was a recent effort in Hadoop to switch from the s3n jets3t-backed FS to "s3a" which is AWS SDK-backed, which claims several benefits over the prior ([HADOOP-10400](https://issues.apache.org/jira/browse/HADOOP-10400)).

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+  }
+
+  @Override
+  public boolean supportsPath(String path, TachyonConf tachyonConf) {
+    return path != null && path.startsWith(Constants.HEADER_S3N);
+  }
+
+  private boolean addAndCheckAWSCredentials(TachyonConf tachyonConf) {
+    String accessKeyConf = "fs.s3n.awsAccessKeyId";
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java]: An alternative way to define the inline string constants accessKeyConf, it could be easier to maintain the code if we have a static class string member that holds this constant.

Similar comment on secreteKeyConf.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import tachyon.conf.TachyonConf;
+import tachyon.underfs.UnderFileSystem;
+import tachyon.underfs.UnderFileSystemFactory;
+
+public class S3UnderFileSystemFactory implements UnderFileSystemFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  @Override
+  public UnderFileSystem create(String path, TachyonConf tachyonConf, Object conf) {
+    Preconditions.checkArgument(path != null, "path may not be null");
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java]: NIT: checkNotNull also?
NIT2: shall we also check-faill if tachyonConf and conf are null?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java (hunk 1 above)
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java]: Actually, just realized that both const strings "fs.s3n.awsSecretAccessKey" and "fs.s3n.awsAccessKeyId" appear multiple times in this package. It make better sense to have them either in tachyon.Constants, or define them in something like tachyon.underfs.s3.Constants.java, and refer them using the constant variable name rather than using inline strings. In that way, mis-spelling in the string content will be caught by compiler/IDE before runtime.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java (hunk 2 above)
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystemFactory.java]: Actually, since 3rd parameter conf is not really used in this constructor, considering naming this param as unusedConf.

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderStorageCluster.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * This class will use Amazon S3 as the backing store. Update awsAccessKey, awsSecretKey, and
+ * baseDirectory accordingly to enable this test.
+ */
+public class S3UnderStorageCluster extends UnderFileSystemCluster {
+
+  private final String awsAccessKey = "";
+  private final String awsSecretKey = "";
+  private final String baseDirectory = ""; // s3n://my-test-bucket/tachyon-test
```
Comment (apc999) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderStorageCluster.java]: I am confused, baseDirectory seems never updated after initialized as an empty string.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3OutputStream.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+public class S3OutputStream extends OutputStream {
+  private final String mBucketName;
+  private final String mKey;
+  private final OutputStream mLocalOutputStream;
+  private final File mFile;
+  private final S3Service mClient;
+
+  public S3OutputStream(String bucketName, String key, S3Service client) throws IOException {
+    mBucketName = bucketName;
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3OutputStream.java]: NIT: is bucketName eligible to be empty? if no, probably have a checkArgument here?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+    if (mkdirs(getParentKey(path), true)) {
+      return new S3OutputStream(mBucketName, stripPrefix(path), mClient);
+    }
+    return null;
+  }
+
+  // Same as create(path)
+  @Override
+  public OutputStream create(String path, int blockSizeByte) throws IOException {
+    return create(path);
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Considering LOG.error (or LOG.warn?) here to indicate blockSizeByte is not used, before return create(path).

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+  public OutputStream create(String path, short replication, int blockSizeByte)
+      throws IOException {
+    return create(path);
+  }
+
+  @Override
+  public boolean delete(String path, boolean recursive) throws IOException {
+    if (!recursive) {
+      if (isFolder(path) && listInternal(path, false).length != 0) {
+        return false;
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Better to LOG.error indicating why delete fails before return false

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+    }
+    // Get all relevant files
+    String[] pathsToDelete = listInternal(path, true);
+    for (String pathToDelete : pathsToDelete) {
+      // If we fail to deleteInternal one file, stop
+      if (!deleteInternal(CommonUtils.concatPath(path, pathToDelete))) {
+        return false;
+      }
+    }
+    deleteInternal(path);
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: return deleteInternal(path) ?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+  // Largest size of one file is 5 TB
+  @Override
+  public long getBlockSizeByte(String path) throws IOException {
+    return Constants.TB * 5;
+  }
+
+  // Not supported
+  @Override
+  public Object getConf() {
+    return null;
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Maybe LOG.error("not supported")?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+    } catch (ServiceException se) {
+      LOG.error("Failed to open file: " + path, se);
+      return null;
+    }
+  }
+
+  @Override
+  public boolean rename(String src, String dst) throws IOException {
+    if (!exists(src)) {
+      return false;
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: better to have LOG.error before return false.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+    }
+    if (exists(dst)) {
+      if (!isFolder(dst)) {
+        LOG.error("Unable to rename " + src + " to " + dst + " because destination already "
+            + "exists as a file.");
+        return false;
+      }
+      String srcName = getKeyName(src);
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: I am confused here. When isFolder(dst), why we should proceed? 
It seems here we copy src under folder dst in this case? Not sure this matches the semantics of rename in hdfs/gluster ufs.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+      mClient.putObject(mBucketName, obj);
+      return true;
+    } catch (ServiceException se) {
+      LOG.error("Failed to create directory: " + key, se);
+      return false;
+    }
+  }
+
+  /**
+   * Strips the s3 bucket prefix from the path if it is present
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Consider add an example here in javadoc.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+   * @param key the key to strip
+   * @return the key without the s3 bucket prefix
+   */
+  private String stripPrefix(String key) {
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: how about stripPrefixIfPresent, to better indicate its function?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+    int separatorIndex = key.lastIndexOf(PATH_SEPARATOR);
+    if (separatorIndex >= 0) {
+      return key.substring(0, separatorIndex);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Determines if the key is represents a folder. If false is returned, it is not guaranteed that
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: delete either "is" or "represents"

**Comment by apc999** [comment]
Made a bunch of comments. Most are tiny readability issues, except:
(1) no real unit-tests on S3UnderFileSystems?
(2) I think the semantics of rename is different from HDFS ufs.  should we be consistent?

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderStorageCluster.java (hunk 3 above)
Comment (calvinjia) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderStorageCluster.java]: The user should update this value if they wish to run s3 integration tests.

**Comment by calvinjia** [comment]
@aarondav Thanks for the info, you bring up a very good point. Originally I thought to use AWS SDK, but I found that the jets3t library / s3n was more widely used and mature. When s3a gains more adoption, it should not be a problem to create another underfs for it; this s3 client is by no means the only option that will exist.

@apc999 Thanks for the review, I've updated the code accordingly.
For the first point, the s3 module will pull in all the tests from core (including the under storage system interface test), which is the same test coverage we currently have for the hdfs. However it is disabled by default because the user will need to use an actual S3 backend. 
For the second point, I've changed it to not support rename to existing folders. The updated code is much simpler!

File: underfs/s3/pom.xml [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>org.tachyonproject</groupId>
+      <artifactId>tachyon</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>net.java.dev.jets3t</groupId>
+      <artifactId>jets3t</artifactId>
+      <version>0.9.3</version>
```
Comment (haoyuan) [inline underfs/s3/pom.xml]: should `jets3t ` be excluded from `hadoop-client` dependency?

File: underfs/s3/pom.xml [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+      <groupId>org.tachyonproject</groupId>
+      <artifactId>tachyon</artifactId>
+      <version>${project.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.thrift</groupId>
+      <artifactId>libthrift</artifactId>
+      <scope>test</scope>
```
Comment (haoyuan) [inline underfs/s3/pom.xml]: why is this needed if this references org.tachyonproject.tachyon?

File: underfs/s3/pom.xml (hunk 14 above)
Comment (calvinjia) [inline underfs/s3/pom.xml]: The direct dependency will take precedence over the transitive dependency, so it shouldn't be necessary.

File: underfs/s3/pom.xml (hunk 15 above)
Comment (calvinjia) [inline underfs/s3/pom.xml]: This was previously included because thrift was shaded in tachyon project, but I think it can be removed now since we changed our shading. Good catch.

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderStorageCluster.java (hunk 3 above)
Comment (apc999) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderStorageCluster.java]: Probably leverage conf to set this value?

Comment (calvinjia) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderStorageCluster.java]: Oh yeah, good call.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+    return deleteInternal(path);
+  }
+
+  @Override
+  public boolean exists(String path) throws IOException {
+    // Root path always exists.
+    return isRoot(path) || getObjectDetails(path) != null;
+  }
+
+  // Largest size of one file is 5 TB
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Could you comment in javadoc why 5TB? is this a requirement from S3? or a limitation in Tachyon?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+  }
+
+  /**
+   * Strips the s3 bucket prefix from the path if it is present. For example, for input key
+   * s3n://my-bucket-name/my-path/file, the output would be my-path/file.
+   * @param key the key to strip
+   * @return the key without the s3 bucket prefix
+   */
+  private String stripPrefixIfPresent(String key) {
+    String prefix = Constants.HEADER_S3N + mBucketName + PATH_SEPARATOR;
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: NIT: this prefix really seems a constant after constructor is called, maybe worth have a private member string to prevent string calucation each time on stripPrefixIfPresent

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 18]
```diff
@@ -0,0 +1,486 @@
...
+                  child.length() - FOLDER_SUFFIX.length()) : child;
+          ret[i] = child;
+        }
+        return ret;
+      }
+      // Non recursive list
+      Set<String> children = new HashSet<String>();
+      for (S3Object obj : objs) {
+        // Remove parent portion of the key
+        String child = obj.getKey().substring(path.length());
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: This logic to derive child name could be a separate function in this class, and I highly recommend to unit test this function. Maybe not for this PR though.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 19]
```diff
@@ -0,0 +1,486 @@
...
+      return null;
+    }
+  }
+
+  /**
+   * Gets the parent key of the input key, or null if no parent exists
+   * @param key the key to get the parent of
+   * @return the the parent key
+   */
+  private String getParentKey(String key) {
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: helper functions like this one and the next should be easily unit-tested, without requiring a S3 mock. Maybe add some unittests (in addition to integration tests) in next PR after unit-tests and integration tests are separated (TACHYON-432)

**Comment by apc999** [comment]
Added a few minor issues.

LGTM :)

Thanks for the PR. this is really an important feature

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 20]
```diff
@@ -0,0 +1,486 @@
...
+import tachyon.conf.TachyonConf;
+import tachyon.underfs.UnderFileSystem;
+import tachyon.util.CommonUtils;
+
+/**
+ * Under file system implementation for S3 using the Jets3t library.
+ */
+public class S3UnderFileSystem extends UnderFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final String FOLDER_SUFFIX = "_$folder$";
```
Comment (haoyuan) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Great to add more comments

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java (hunk 16 above)
Comment (haoyuan) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: It's great to change this to be java doc
